Components returning null with different results for render and shallow

I came across an interesting inconsistency, let me demonstrate this with some code, here's a simplified version of my `ContentToggle` component:

``` js
export default class ContentToggle extends React.Component {
  props: {
    component: Function|string,
    content: ?ReactElement,
    options: ?Array<ReactElement>,
  };

  render(): ?ReactElement {
    const {content, options, ...otherProps} = this.props;

    if (!content && !options) {
      return null;
    }

    return React.createElement('a', { href: '#', ...otherProps });
  }
}
```

Important to note is that I either need to pass on `content` or `options` as props or otherwise it will return `null`.

My initial tests I wrote:

``` js
describe(ContentToggle.displayName, () => {
  it('renders children if content is passed', () => {
    const children = 'FOO';
    const el = render(<ContentToggle content={<span />}>{children}</ContentToggle>);

    expect(el).to.have.text(children);
  });

  it('renders children if options are passed', () => {
    const children = 'FOO';
    const el = render(<ContentToggle options={[]}>{children}</ContentToggle>);

    expect(el).to.have.text(children);
  });

 it('renders nothing if no content or options are passed', () => {
    const children = 'FOO';
    const el = render(<ContentToggle>{children}</ContentToggle>);

    expect(el).to.not.have.text(children);
    expect(el).to.be.blank();
  });
});
```

The first two tests passed, but the third one did not:

```
  1) undefined renders nothing if no content or options are passed:
     AssertionError: expected the node in <??? /> to be empty

     HTML:

     <noscript></noscript>
```

When I change the 3rd test to use `shallow` vs `render` this test passes, or I can write the test as follow:

``` js
describe(ContentToggle.displayName, () => {
  it('renders nothing if no content or options are passed', () => {
    const children = 'FOO';
    const el = render(<ContentToggle>{children}</ContentToggle>);

    expect(el).to.not.have.text(children);
    expect(el.find('noscript')).to.be.blank();
  });
});
```

I would have expected `expect(el).to.be.blank();` to work with both `render` as with `shallow` (or `mount` for that matter), or at least consistent behaviour across all 3.

After discussing this in https://github.com/airbnb/enzyme/issues/139 we've decided to tackle this within `chai-enzyme`.

--- a/README.md
+++ b/README.md
@@ -307,6 +307,17 @@ expect(wrapper.find('#parent')).to.not.be.blank()
 
 expect(wrapper.find('#child')).to.be.empty // an alias
 expect(wrapper.find('#parent')).to.not.be.empty // an alias
+
+class NullFixture extends React.Component {
+  render () {
+    return null
+  }
+}
+
+const nullWrapper = moun(<NullFixture />) // mount/render/shallow when applicable
+
+expect(wrapper).to.be.blank()
+expect(wrapper).to.be.empty // an alias
 ```
 
 #### `present()`
@@ -334,6 +345,17 @@ const wrapper = mount(<Fixture />) // mount/render/shallow when applicable
 
 expect(wrapper.find('#parent')).be.present()
 expect(wrapper.find('#parent')).to.exist // an alias
+
+class NullFixture extends React.Component {
+  render () {
+    return null
+  }
+}
+
+const nullWrapper = moun(<NullFixture />) // mount/render/shallow when applicable
+
+expect(wrapper).to.be.present()
+expect(wrapper).to.exist // an alias
 ```
 
 #### `html(str)`

--- a/README.md
+++ b/README.md
@@ -314,7 +314,7 @@ class NullFixture extends React.Component {
   }
 }
 
-const nullWrapper = moun(<NullFixture />) // mount/render/shallow when applicable
+const nullWrapper = mount(<NullFixture />) // mount/render/shallow when applicable
 
 expect(wrapper).to.be.blank()
 expect(wrapper).to.be.empty // an alias
@@ -352,7 +352,7 @@ class NullFixture extends React.Component {
   }
 }
 
-const nullWrapper = moun(<NullFixture />) // mount/render/shallow when applicable
+const nullWrapper = mount(<NullFixture />) // mount/render/shallow when applicable
 
 expect(wrapper).to.be.present()
 expect(wrapper).to.exist // an alias

--- a/src/CheerioTestWrapper.js
+++ b/src/CheerioTestWrapper.js
@@ -16,6 +16,14 @@ export default class CheerioTestWrapper extends TestWrapper {
     return 'the node in <??? />'
   }
 
+  isEmpty () {
+    if (this.tagName() === 'noscript') {
+      return true
+    }
+
+    return super.isEmpty()
+  }
+
   attr (name) {
     return this.el.attr(name)
   }

--- a/test/null.test.js
+++ b/test/null.test.js
@@ -1,0 +1,35 @@
+class Fixture extends React.Component {
+  render () {
+    return null
+  }
+}
+
+const it = createTest(<Fixture />)
+
+describe('#blank', () => {
+  describe('()', () => {
+    it('passes when the actual matches the expected', (wrapper) => {
+      expect(wrapper).to.be.blank()
+    })
+
+    it('fails when the actual does not match the expected', (wrapper) => {
+      expect(() => {
+        expect(wrapper).to.not.be.blank()
+      }).to.throw(`not to be empty`)
+    })
+  })
+})
+
+describe('#present', () => {
+  describe('()', () => {
+    it('passes when the actual matches the expected', (wrapper) => {
+      expect(wrapper).to.be.present()
+    })
+
+    it('fails when the actual does not match the expected', (wrapper) => {
+      expect(() => {
+        expect(wrapper).to.not.be.present()
+      }).to.throw(`not to exist`)
+    })
+  })
+})

